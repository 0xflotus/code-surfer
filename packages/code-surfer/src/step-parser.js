const mapRange = range => {
  if (!range) return {};
  const newTokens = {};
  const [start, end] = range;
  for (let i = start; i <= end; i++) {
    newTokens[i] = null;
  }
  return newTokens;
};
const mapLines = lines => {
  const newTokens = {};
  lines.forEach(line => (newTokens[line] = null));
  return newTokens;
};

const getTokensPerLineFromObject = ({
  lines = [],
  range,
  ranges = [],
  tokens
}) => {
  const newTokens = {};

  Object.assign(newTokens, mapLines(lines));
  Object.assign(newTokens, mapRange(range));
  Object.assign(newTokens, ...ranges.map(mapRange));
  Object.assign(newTokens, tokens);

  return newTokens;
};

const expandString = part => {
  // Transforms something like
  // - "1:3" to [1,2,3]
  // - "4" to [4]
  const [start, end] = part.split(":");
  if (!end) {
    return [+start];
  }
  const list = [];
  for (let i = +start; i <= +end; i++) {
    list.push(i);
  }
  return list;
};

const getTokensPerLineFromString = step => {
  const parts = step.split(/,(?![^\[]*\])/g).map(part => {
    const tokensMatch = part.match(/(\d+)\[(.+)\]/);
    if (tokensMatch) {
      const [_, line, tokens] = tokensMatch;
      const tokenList = tokens.split(",").map(expandString);
      return { [line]: [].concat(...tokenList) };
    }

    const [start, end] = part.split(":");
    return mapRange([+start, +end || +start]);
  });

  return Object.assign({}, ...parts);
};

const getTokensPerLine = step => {
  if (typeof step === "string") {
    return getTokensPerLineFromString(step);
  } else {
    return getTokensPerLineFromObject(step);
  }
};

export default getTokensPerLine;
