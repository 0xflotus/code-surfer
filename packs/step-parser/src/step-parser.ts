import { linesDiff } from "./differ";
import { tokenize } from "./tokenizer";
import { parseFocus } from "./focus-parser";
import { toEntries } from "./object-entries";
import { applyPatch } from "diff";

export function parseSteps(
  inputSteps: { code: string; focus?: string; lang?: string }[]
) {
  if (inputSteps.length === 0) {
    return {
      tokens: [],
      types: [],
      steps: []
    };
  }

  const lang = inputSteps[0].lang;

  if (!lang) {
    throw new Error("Missing code language");
  }

  const codeList = getCodeList(inputSteps);

  const { lineIds, steps } = linesDiff(codeList);
  const allTokens: string[][] = [];
  const allTypes: string[][] = [];
  const allSteps: {
    lines: number[];
    focus: Record<number, true | number[]>;
  }[] = [];

  steps.forEach((step, i) => {
    const code = codeList[i];
    const { tokens, types } = tokenize(code, lang);
    const lineKeys: number[] = [];
    step.forEach((lineId, lineIndex) => {
      const lineKey = lineIds.indexOf(lineId);
      allTokens[lineKey] = tokens[lineIndex];
      allTypes[lineKey] = types[lineIndex];
      lineKeys.push(lineKey);
    });

    const focusString = inputSteps[i].focus;
    const prevLineKeys = allSteps[i - 1] ? allSteps[i - 1].lines : [];
    const focus = focusString
      ? parseFocus(focusString)
      : getDefaultFocus(prevLineKeys, lineKeys);
    allSteps.push({ lines: lineKeys, focus });
  });

  // split tokens into columns when needed
  allSteps.forEach(({ lines, focus }) => {
    toEntries(focus).forEach(([lineIndex, lineFocus]) => {
      if (Array.isArray(lineFocus)) {
        const lineKey = lines[lineIndex];
        const lineTypes = allTypes[lineKey];
        const lineTokens = allTokens[lineKey];
        const [newTypes, newTokens] = splitIntoColumns(lineTypes, lineTokens);
        allTypes[lineKey] = newTypes;
        allTokens[lineKey] = newTokens;
      }
    });
  });

  return {
    tokens: allTokens,
    types: allTypes,
    steps: allSteps
  };
}

function getCodeList(
  inputSteps: {
    code: string;
    lang?: string | undefined;
  }[]
) {
  const firstLang = inputSteps[0].lang;
  if (firstLang === "diff") {
    return inputSteps.map(s => s.code);
  }

  let prevCode = "";
  return inputSteps.map(({ code, lang }) => {
    let stepCode = lang === "diff" ? applyPatch(prevCode, code) : code;
    prevCode = stepCode;
    return stepCode;
  });
}

function splitIntoColumns(
  types: string[],
  tokens: string[]
): [string[], string[]] {
  const newTypes: string[] = [];
  const newTokens: string[] = [];
  tokens.forEach((token, i) => {
    const tokenType = types[i];
    Array.from(token).forEach(char => {
      newTokens.push(char);
      newTypes.push(tokenType);
    });
  });
  return [newTypes, newTokens];
}

function getDefaultFocus(prevLineKeys: number[], lineKeys: number[]) {
  const focus = {} as Record<number, true>;
  lineKeys.forEach((lineKey, lineIndex) => {
    if (!prevLineKeys.includes(lineKey)) {
      focus[lineIndex] = true;
    }
  });

  if (Object.keys(focus).length === 0) {
    lineKeys.forEach((_, lineIndex) => {
      focus[lineIndex] = true;
    });
  }

  return focus;
}
