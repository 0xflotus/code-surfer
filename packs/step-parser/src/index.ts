import { parseSteps } from "step-parser";

export default parseSteps;

export function flat<T>(nestedArrays: T[][]) {
  return nestedArrays.flat();
}
