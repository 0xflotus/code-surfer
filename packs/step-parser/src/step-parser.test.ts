import { parseSteps } from "./step-parser";

describe("Parsing steps", () => {
  it("works", () => {
    const steps = [
      {
        code: `
console.log(1)
console.log(2)
console.log(3)
      `.trim(),
        lang: "javascript",
        focus: "1,2"
      },
      {
        code: `
console.log(1)
console.log(3)
      `.trim(),
        lang: "javascript",
        focus: "1[3:5],2"
      }
    ];
    const result = parseSteps(steps);
    expect(result).toMatchSnapshot();
  });

  it("adds default focus", () => {
    const steps = [
      {
        code: `
console.log(1)
console.log(2)
console.log(3)
      `.trim(),
        lang: "javascript"
      },
      {
        code: `
console.log(1)
console.log(3)
console.log(4)
      `.trim()
      }
    ];
    const result = parseSteps(steps);
    expect(result).toMatchSnapshot();
  });

  it.only("works with empty diff", () => {
    const steps = [
      {
        code: `
console.log(1)
console.log(2)
console.log(3)
      `.trim(),
        lang: "javascript"
      },
      {
        code: ``,
        lang: "diff"
      }
    ];
    const result = parseSteps(steps);
    expect(result).toMatchSnapshot();
  });
});
