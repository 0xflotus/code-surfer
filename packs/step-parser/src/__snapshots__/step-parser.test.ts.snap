// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsing steps adds default focus 1`] = `
Object {
  "steps": Array [
    Object {
      "focus": Object {
        "0": true,
        "1": true,
        "2": true,
      },
      "focusCenter": 0.5,
      "focusCount": 1,
      "lines": Array [
        0,
        1,
        2,
      ],
      "longestLineIndex": 0,
    },
    Object {
      "focus": Object {
        "2": true,
        "3": true,
      },
      "focusCenter": 2.5,
      "focusCount": 1,
      "lines": Array [
        0,
        2,
        3,
        4,
      ],
      "longestLineIndex": 3,
    },
  ],
  "tokens": Array [
    Array [
      "console",
      ".",
      "log",
      "(",
      "1",
      ")",
    ],
    Array [
      "console",
      ".",
      "log",
      "(",
      "2",
      ")",
    ],
    Array [
      "console",
      ".",
      "log",
      "(",
      "3",
      ")",
    ],
    Array [],
    Array [
      "console",
      ".",
      "log",
      "(",
      "4",
      ")",
      ";",
    ],
  ],
  "types": Array [
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
    Array [],
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
      "punctuation",
    ],
  ],
}
`;

exports[`Parsing steps works 1`] = `
Object {
  "steps": Array [
    Object {
      "focus": Object {
        "0": true,
        "1": true,
      },
      "focusCenter": 0.5,
      "focusCount": 1,
      "lines": Array [
        0,
        1,
        2,
      ],
      "longestLineIndex": 0,
    },
    Object {
      "focus": Object {
        "0": Array [
          2,
          3,
          4,
        ],
        "1": true,
      },
      "focusCenter": 0.5,
      "focusCount": 1,
      "lines": Array [
        0,
        2,
      ],
      "longestLineIndex": 0,
    },
  ],
  "tokens": Array [
    Array [
      "c",
      "o",
      "n",
      "s",
      "o",
      "l",
      "e",
      ".",
      "l",
      "o",
      "g",
      "(",
      "1",
      ")",
    ],
    Array [
      "console",
      ".",
      "log",
      "(",
      "2",
      ")",
    ],
    Array [
      "console",
      ".",
      "log",
      "(",
      "3",
      ")",
    ],
  ],
  "types": Array [
    Array [
      "plain",
      "plain",
      "plain",
      "plain",
      "plain",
      "plain",
      "plain",
      "punctuation",
      "function",
      "function",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
  ],
}
`;

exports[`Parsing steps works with empty diff 1`] = `
Object {
  "steps": Array [
    Object {
      "focus": Object {
        "0": true,
        "1": true,
        "2": true,
      },
      "focusCenter": 0.5,
      "focusCount": 1,
      "lines": Array [
        0,
        1,
        2,
      ],
      "longestLineIndex": 0,
    },
    Object {
      "focus": Object {
        "0": true,
        "1": true,
        "2": true,
      },
      "focusCenter": 0.5,
      "focusCount": 1,
      "lines": Array [
        0,
        1,
        2,
      ],
      "longestLineIndex": 0,
    },
  ],
  "tokens": Array [
    Array [
      "console",
      ".",
      "log",
      "(",
      "1",
      ")",
    ],
    Array [
      "console",
      ".",
      "log",
      "(",
      "2",
      ")",
    ],
    Array [
      "console",
      ".",
      "log",
      "(",
      "3",
      ")",
    ],
  ],
  "types": Array [
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
    Array [
      "plain",
      "punctuation",
      "function",
      "punctuation",
      "number",
      "punctuation",
    ],
  ],
}
`;
